README
tank

A real-time tank-shooting game with two tanks, for SDL.


Layers (8):
0 - stars & sun & moon
1 - land & water
2 - clouds
3 - rain
5 - tanks
6 - shots and lightning
7 - arrows, text

Author: Martin Hradil <himdel seznam cz>
Licence: GNU GPLv2 or later


ChangeLog
2010
	* merged 0.8a, 0.8e and a 0.8a fork, but not the 0.8d debug stuff, some changes here are not in the code
	* this ChangLog is only historical now, for new changes, see git log


2005-08-28 18:35:48:	himdel
	* tank-0.8e released
	* added scripts/tgz.sh

2005-08-27 00:04:34:	himdel
//	* fixed the debug trace system:
//	* some functions didnt call debug_add but did _out, so we used to forget where we are
//	* changed the format of trace dump to func per line and added number
//	* added warnings about debug_out&add count mismatch

2005-08-26 21:46:42:	himdel
//	* -DDEBUG no longer in Makefiles but in debug.h
//	* that way you controll everything by just including debug.h or 

2005-08-23 19:38:23:	himdel
//	* tank-0.8d released - a debug release
//	@ debug.[ch] added
//	* -DDEBUG in Makefiles
//	* each func has debug_add(NAME) and debug_out - don't forget to call BEFORE return

2005-08-23 17:58:58:	himdel
	* umm, another SDL problem. I'm stupid and don't know how to handle strings. Sorry.

2005-08-23 14:54:28:	himdel
	* tank-0.8c released (SDL keyboard fix version)

2005-08-22 21:00:22:	himdel
	* added tank08B.png to contrib	(my thanks to Petr Schreiber!)
	* also recieved a bugreport -> most of it follows as solved bugs :)
	@ removed unused timezone from paint.c - didn't compile under gcc4 --std=c99
	* all keys should work now under SDL too (bug in h__sk2SK)
//	* moved images to pic/ => @ him_showimg assumes pic/, Makefile.wine too

2005-08-20 22:44:30:	himdel
//	* minor corrections to win/Makefile.wine and win/zip.sh
	* when changing version number, do it just in Makefile and README and add an entry to ChangeLog

2005-08-20 22:08:39:	himdel
	* tank-0.8b released
//	* added win/Makefile.wine to compile using wine
//	* winzip target to Makefile

2005-08-20 21:25:08:	himdel
	* define to enable land2 added
	* land2 generator works ok now
	* struct playa p[2] is global now => changed headers of functions that use it
	* keys: p0 (rdfg TAB), p1 (arrows 0)
	* killed dostuff - moved to main loop

2005-08-19 23:20:56:	himdel
	* colored arrows, score and energy bars

2005-08-19 23:19:27:	himdel
	* windows related fix: moved echo from Makefile to genload.pl
	* changed version number to tank-0.8b


2005-08-19 23:09:27:	himdel
	* tank-0.8a released

2005-08-19 21:01:43:	himdel
	* land slides no longer buggy (used to duplicate land)
	* land is heavier than water
	* generating terrain notice
	* weapon energy
	@ dostuff rewritten to use struct Tasks * table and run things every # ms

2005-08-01 10:52:47:	himdel
	* land slides sidewise too now
	@ added unsigned int tm into strcut lndpts
	@ added him_getnow() into paint (in SDL uses GetTicks, in vga uses gettimeofday(2))
	* land is more random after explosion


0.6 -> 0.7	(07/14/2004)

main - eliminated the "explosion on the border of screen hangs up the
       whole game" bug
     - added real-looking (almost) lightning
     - land slides down

water - water behaves (ALMOST - see README and grep BUG) properly now

paint - water and clouds are transparent (~50%) when there's sun or moon behind

paintSDL - uses SDL_Surfaces for layers
    


0.5 -> 0.6	(once)

paint - added him_ufullcircler
main - realtime shooting (, changed keys, multiple explosions)

moved raw keyboard to paint*

paintSDL - created => tankSDL


menu.c
/*
  menu.c - menu for tank
  himdel@seznam.cz
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "paint.h"
#include "options.h"
#include "letters.h"
#include "menu.h"
#include "gui.h"

int menu_p0 = 0;
int menu_p1 = 0;

int mnu_settings ();
int mnu_network ();



int
startmenu (ait)
  int ait;
{
  int foo;

  him_clrscr ();  
  
  him_putlock ();
  
  if (him_showimg ("img/menu.png") >= 0)
    {
      int x, y;
      
      him_setpalette (0, 0, 0, 0);
      him_setpalette (4, 0xa8, 0, 0);
      him_setpalette (12, 0xfc, 0x54, 0x54);
      
      for (x = 256; x < 384; x++)
        for (y = 160; y < 320; y++)
          if ((x + y) % 2)
            him_putpixel (x, y, 0);
      
      wrtwrd (320 - 8, 172, "1v1", 0, -2, -1);
      wrtwrd (320 - 12, 204, "1vPC", 0, -2, -1);
      wrtwrd (320 - 16, 236, "1vNET", 0, -2, -1);
      wrtwrd (320 - 28, 268, "settings", 0, -2, -1);
      wrtwrd (320 - 12, 300, "quit", 0, -2, -1);

      wrtwrd (320 - 14, 166, "1v1", 0, -2, -1);
      wrtwrd (320 - 18, 198, "1vPC", 0, -2, -1);
      wrtwrd (320 - 22, 230, "1vNET", 0, -2, -1);
      wrtwrd (320 - 34, 262, "settings", 0, -2, -1);
      wrtwrd (320 - 18, 294, "quit", 0, -2, -1);

      him_filledtriangle (256, 168, 256, 168 + 16, 268, 168 + 8, 0, -1);
      him_filledtriangle (256, 200, 256, 200 + 16, 268, 200 + 8, 0, -1);
      him_filledtriangle (256, 232, 256, 232 + 16, 268, 232 + 8, 0, -1);
      him_filledtriangle (256, 264, 256, 264 + 16, 268, 264 + 8, 0, -1);
      him_filledtriangle (256, 296, 256, 296 + 16, 268, 296 + 8, 0, -1);
    }

  wrtwrd (320 - 10, 170, "1v1", 4, -2, -1);
  wrtwrd (320 - 14, 202, "1vPC", 4, -2, -1);
  wrtwrd (320 - 18, 234, "1vNET", 4, -2, -1);
  wrtwrd (320 - 30, 266, "settings", 4, -2, -1);
  wrtwrd (320 - 14, 298, "quit", 4, -2, -1);
  
  wrtwrd (320 - 12, 168, "1v1", 12, -2, -1);
  wrtwrd (320 - 16, 200, "1vPC", 12, -2, -1);
  if (him_is_SDL)
    wrtwrd (320 - 20, 232, "1vNET", 12, -2, -1);
  wrtwrd (320 - 32, 264, "settings", 12, -2, -1);
  wrtwrd (320 - 16, 296, "quit", 12, -2, -1);

  foo = 168 + (((ait + 4) % 5) * 32); 
  him_filledtriangle (256, foo, 256, foo + 16, 268, foo + 8, 12, -1);

  him_putulock ();
  him_putupd ();

  while (1)
    {
      static int u = 0, d = 0;  /* up, down */

      him_keyupd ();
      
      if (him_keypr (SCANCODE_ESCAPE))
        return 0;
      
      if (him_keypr (SCANCODE_ENTER))
        {
          do
            {
              him_keyupd ();
            }
          while (him_keypr (SCANCODE_ENTER));

          switch (ait)
            {
              case 0:
                return 0;
                break;
              case 1:
                menu_p0 = 0;
                menu_p1 = 0;
                return 1;
                break;
              case 2:
                menu_p0 = 0;
                menu_p1 = 2;
                return 2;
                break;
              case 3:
                if ((foo = mnu_network ()))
                  {
                    menu_p0 = menu_p1 = 0;
                    if (foo == 1)
                      menu_p0 = 1;
                    else if (foo == 2)
                      menu_p1 = 1;
                    else
                      {
                        printf ("guru medication 0x01\n");
                        return 0;
                      }
                    return 3;
                  }
                break;
              case 4:
                if (mnu_settings ())
                  opt_save ();
                break;
              default:
                printf ("guru medication 0x00\n");
                return 0;
            }
        }
      
      if (d && (!him_keypr (SCANCODE_CURSORBLOCKDOWN)))
        d = 0;
      if (u && (!him_keypr (SCANCODE_CURSORBLOCKUP)))
        u = 0;
      
      if ((d == 0) && (him_keypr (SCANCODE_CURSORBLOCKDOWN)))
        {
          him_putlock ();
          
          d = 1;
          
          foo = 168 + (((ait + 4) % 5) * 32); 
          him_filledtriangle (256, foo, 256, foo + 16, 268, foo + 8, 0, -1);

          ait = ((ait + 1) % 5);    
          if ((him_is_SDL == 0) && (ait == 3))
            ait = ((ait + 1) % 5);

          foo = 168 + (((ait + 4) % 5) * 32); 
          him_filledtriangle (256, foo, 256, foo + 16, 268, foo + 8, 12, -1);
 
          him_putulock ();
          him_putupd ();
        }

      if ((u == 0) && (him_keypr (SCANCODE_CURSORBLOCKUP)))
        {
          him_putlock ();

          u = 1;
          
          foo = 168 + (((ait + 4) % 5) * 32); 
          him_filledtriangle (256, foo, 256, foo + 16, 268, foo + 8, 0, -1);
          
          ait = ((ait + 4) % 5);
          if ((him_is_SDL == 0) && (ait == 3))
            ait = ((ait + 4) % 5);

          foo = 168 + (((ait + 4) % 5) * 32); 
          him_filledtriangle (256, foo, 256, foo + 16, 268, foo + 8, 12, -1);

          him_putulock ();
          him_putupd ();
        }
    }

  return 0;
}



int
mnu_network ()
{
  int x, y;
  
  him_putlock ();
  him_showimg ("img/netgame.png");
      
  for (x = 200; x < 440; x++)
    for (y = 212; y < 272; y++)
      him_putpixel (x, y, 0);
      
  wrtwrd (208, 216, "connect to: ", 2, -2, -1);
  wrtwrd (208, 248, "be server", 1, -2, -1);

  him_box (304, 215, 432, 232, 1, -1);
  him_box (305, 216, 431, 231, 0, -1);

  him_putulock ();
  him_putupd ();

/* TODO: act as edit ten ramecek - zadas IP, ta se da do SDLNet_ResolveHost, ... */
//  SDLNet_Init ();

sleep (5);
//  SDLNet_Quit ();

  return 0;
}


int
mnu_settings ()
{
  printf ("TODO: mnu_settings ()\n");
  return 0;
}
